Sources/SecureXPC/Server/XPCServer.swift

```diff
@@ -340,54 +310,88 @@ public class XPCServer {
 // These wrappers perform type erasure via their implemented protocols while internally maintaining type constraints
 // This makes it possible to create heterogenous collections of them
 
-fileprivate protocol XPCHandlerWithoutMessageWithoutReply {
-    func handle() throws -> Void
+fileprivate protocol XPCHandler {
+    func handle(request: Request, reply: inout xpc_object_t?) throws
 }
 
-fileprivate struct ConstrainedXPCHandlerWithoutMessageWithoutReply: XPCHandlerWithoutMessageWithoutReply {
+fileprivate struct ConstrainedXPCHandlerWithoutMessageWithoutReply: XPCHandler {
     let handler: () throws -> Void
     
-    func handle() throws {
+    func handle(request: Request, reply: inout xpc_object_t?) throws {
```

Comment:
The approach you took here causes some duplication, but it has an opportunity to make more specific error messages.

If you're curious to compare notes, checkout what I came up with here: https://github.com/trilemma-dev/SecureXPC/commit/96d9c92430592a21f748d8767b58536a7ecd1ab9

Reply:
Ah I didn't realize you were working on this, sorry if I've caused wasted effort.

Looking at your approach, they're certainly similar although I prefer the approach I took a bit as it generates more detailed error messages and doesn't have any forced casts.

Reply:
> Ah I didn't realize you were working on this, sorry if I've caused wasted effort.

That's kind of on me, I should have communicated that. In general, I have quite a bit more goodies built on top of this framework in my private repo, but it's generally a bunch of hacks and not-yet-shippable code.

I'm working to upstream them as fast as I can haha.

> I prefer the approach I took a bit as it generates more detailed error messages

Indeed, I agree this is nicer. I was apprehensive to do it just because I was too lazy to code out all the cases :D

One critique I'll make however is that this has a "one error at a time" quality to it, which can be frustrating as a user. You run your code, get an error, fix it, run it again, get another error, and so on. It would be nice to have the full error (either your message type is wrong, or your reply type, *or perhaps both*) in one shot.

> doesn't have any forced casts.

If you're new to Swift (though I doubt it, because you write pretty damn good code), I'd caution against always riding the "`try!`/`!`/`as!` is bad" train to its terminal station. I understand the appeal, but there's times (like this one) where I think it can be appropriate to take these escape hatches.

The force casts simply assert "I can't express this type constraint in Swift's type system", but it doesn't mean "I didn't type-check this", because there's type checking *of a different kind*. I literally called my function "typecheck" 😄

Reply:
> One critique I'll make however is that this has a "one error at a time" quality to it, which can be frustrating as a user. You run your code, get an error, fix it, run it again, get another error, and so on. It would be nice to have the full error (either your message type is wrong, or your reply type, or perhaps both) in one shot.

I'll centralize the error generation and then aggregate the error messages so they can be appended to one another.

> I'd caution against always riding the "try!/!/as! is bad" train to its terminal station. 

I agree with this sentiment, we're likely just at different points along the track. I find it valuable to within reason maximize the amount of issues the compiler can catch for me, so I try to avoid `as!` and `!` in many circumstances for that reason. 

My views on `try!` are quite different. I'm working on a service that needs very high reliability with structured error reporting back to a server in case of failure, so to simply crash the process when something goes wrong would be counterproductive - that's likely influencing the approach I'm taking here.

Reply:
I see, that makes perfect sense!